File: schism-android/app/src/main/java/ai/schism/split/core/net/Dto.kt
package ai.schism.split.core.net

import kotlinx.serialization.Serializable

// DTOs mirror schism-backend/docs/api-contract.md (camelCase JSON, money as Long minor units).

@Serializable
data class ParticipantDto(
    val id: String = "",
    val groupId: String = "",
    val name: String = "",
    val userId: String? = null,
)

@Serializable
data class UserDto(
    val id: String,
    val name: String = "",
    val email: String = "",
    val phone: String = "",
    val createdAt: String = "",
    /** Bearer token, present only in the register response; stored and sent on later requests. */
    val token: String? = null,
)

@Serializable
data class UserRequest(
    val name: String,
    val email: String = "",
    val phone: String = "",
)

@Serializable
data class GroupDto(
    val id: String,
    val name: String,
    val information: String = "",
    val currency: String = "",
    val currencyCode: String = "",
    val createdAt: String = "",
    val participants: List<ParticipantDto> = emptyList(),
)

@Serializable
data class CategoryDto(
    val id: Int,
    val grouping: String,
    val name: String,
)

@Serializable
data class PaidForDto(
    val participantId: String,
    val shares: Long,
)

@Serializable
data class ExpenseDto(
    val id: String,
    val groupId: String = "",
    val title: String,
    val amount: Long,
    val categoryId: Int = 0,
    val expenseDate: String = "",
    val paidById: String,
    val splitMode: String = "EVENLY",
    val isReimbursement: Boolean = false,
    val notes: String = "",
    val createdAt: String = "",
    val addedBy: String = "",
    val paidFor: List<PaidForDto> = emptyList(),
)

@Serializable
data class BalanceDto(
    val paid: Long = 0,
    val paidFor: Long = 0,
    val total: Long = 0,
)

@Serializable
data class ReimbursementDto(
    val from: String,
    val to: String,
    val amount: Long,
)

@Serializable
data class BalancesResponseDto(
    val balances: Map<String, BalanceDto> = emptyMap(),
    val reimbursements: List<ReimbursementDto> = emptyList(),
)

@Serializable
data class ActivityDto(
    val id: String,
    val groupId: String = "",
    val time: String = "",
    val activityType: String = "",
    val participantId: String? = null,
    val expenseId: String? = null,
    val data: String = "",
    val amount: Long? = null,
)

// ---- requests ----

@Serializable
data class AuthRequest(
    val email: String,
    val password: String,
    val name: String = "",
    val phone: String = "",
)

@Serializable
data class MyGroupsDto(
    val groupIds: List<String> = emptyList(),
)

@Serializable
data class AuthResponse(
    val id: String,
    val name: String = "",
    val email: String = "",
    val token: String = "",
)

@Serializable
data class ParticipantRequest(
    val id: String? = null,
    val name: String,
    val userId: String? = null,
    /** Phone number for invite-by-contact: lets the backend link this participant when they join. */
    val phone: String? = null,
)

@Serializable
data class CreateGroupRequest(
    val name: String,
    val information: String = "",
    val currency: String,
    val currencyCode: String = "",
    val participants: List<ParticipantRequest>,
)

@Serializable
data class CreateGroupResponse(
    val groupId: String,
)

@Serializable
data class ExpenseRequest(
    val title: String,
    val amount: Long,
    val categoryId: Int = 0,
    val expenseDate: String? = null,
    val paidById: String,
    val splitMode: String = "EVENLY",
    val isReimbursement: Boolean = false,
    val notes: String = "",
    val addedBy: String? = null,
    val paidFor: List<PaidForDto>,
)

// ---- dashboards ----

@Serializable
data class CategoryTotalDto(
    val categoryId: Int,
    val grouping: String = "",
    val name: String = "",
    val amount: Long = 0,
    val count: Int = 0,
)

@Serializable
data class ParticipantTotalDto(
    val participantId: String,
    val name: String = "",
    val paid: Long = 0,
    val share: Long = 0,
    val net: Long = 0,
)

@Serializable
data class MonthTotalDto(
    val month: String,
    val amount: Long = 0,
    val count: Int = 0,
)

@Serializable
data class ExpenseSummaryDto(
    val id: String,
    val title: String,
    val amount: Long,
    val date: String = "",
    val paidById: String = "",
    val categoryId: Int = 0,
)

@Serializable
data class PersonalInGroupDto(
    val participantId: String,
    val name: String = "",
    val paid: Long = 0,
    val share: Long = 0,
    val net: Long = 0,
    val expenseCount: Int = 0,
    val byCategory: List<CategoryTotalDto> = emptyList(),
)

@Serializable
data class GroupDashboardDto(
    val groupId: String,
    val name: String = "",
    val currency: String = "",
    val currencyCode: String = "",
    val totalSpending: Long = 0,
    val expenseCount: Int = 0,
    val reimbursementCount: Int = 0,
    val averageExpense: Long = 0,
    val firstExpenseDate: String? = null,
    val lastExpenseDate: String? = null,
    val byCategory: List<CategoryTotalDto> = emptyList(),
    val byParticipant: List<ParticipantTotalDto> = emptyList(),
    val byMonth: List<MonthTotalDto> = emptyList(),
    val topExpenses: List<ExpenseSummaryDto> = emptyList(),
    val personal: PersonalInGroupDto? = null,
)

@Serializable
data class PersonalGroupSliceDto(
    val groupId: String,
    val groupName: String = "",
    val currency: String = "",
    val currencyCode: String = "",
    val participantId: String = "",
    val participantName: String = "",
    val paid: Long = 0,
    val share: Long = 0,
    val net: Long = 0,
    val expenseCount: Int = 0,
)

@Serializable
data class CurrencyTotalDto(
    val currencyCode: String = "",
    val currency: String = "",
    val paid: Long = 0,
    val share: Long = 0,
    val net: Long = 0,
    val groupCount: Int = 0,
)

@Serializable
data class PersonalDashboardDto(
    val identity: String = "",
    val groupCount: Int = 0,
    val groups: List<PersonalGroupSliceDto> = emptyList(),
    val totals: List<CurrencyTotalDto> = emptyList(),
)

// ---- claim links (alpha) ----

@Serializable
data class ClaimItemDto(
    val idx: Int,
    val name: String,
    val qty: Int,
    val amountMinor: Long,
)

@Serializable
data class ClaimDto(
    val itemIdx: Int,
    val participantId: String,
    val weight: Double,
)

/** One labelled tax/charge line on a claim session, e.g. `{label: "SGST 2.5%", amountMinor: 2360}`. */
@Serializable
data class TaxLineDto(
    val label: String,
    val amountMinor: Long,
)

@Serializable
data class ClaimSessionDto(
    val id: String,
    val groupId: String = "",
    val creatorParticipantId: String = "",
    val title: String = "",
    val currency: String = "",
    val status: String = "open",
    val items: List<ClaimItemDto> = emptyList(),
    val taxMinor: Long = 0,
    val feesMinor: Long = 0,
    val discountMinor: Long = 0,
    val roundoffMinor: Long = 0,
    val version: Int = 1,
    val expenseId: String? = null,
    val claims: List<ClaimDto> = emptyList(),
    val owesPreview: Map<String, Long> = emptyMap(),
    /** Participant ids who marked themselves "done" claiming (advisory, not a finalize gate). */
    val readyParticipantIds: List<String> = emptyList(),
    /** Labelled breakdown of taxMinor (e.g. SGST + CGST), empty when the session has none. */
    val taxes: List<TaxLineDto> = emptyList(),
)

@Serializable
data class CreateClaimSessionRequest(
    val title: String,
    val currency: String,
    val items: List<ClaimItemDto>,
    val taxMinor: Long = 0,
    val feesMinor: Long = 0,
    val discountMinor: Long = 0,
    val roundoffMinor: Long = 0,
    val taxes: List<TaxLineDto> = emptyList(),
)

@Serializable
data class ClaimWeightDto(
    val itemIdx: Int,
    val weight: Double,
)

@Serializable
data class PutClaimsRequest(
    val expectedVersion: Int,
    val weights: List<ClaimWeightDto>,
)

@Serializable
data class ResolutionDto(
    val itemIdx: Int,
    val mode: String,
    val participantId: String? = null,
)

@Serializable
data class FinalizeRequest(
    val expectedVersion: Int,
    val resolutions: List<ResolutionDto>,
)

@Serializable
data class FinalizeResponse(
    val expenseId: String,
)

@Serializable
data class SetReadyRequest(
    val ready: Boolean,
)

@Serializable
data class EditItemsRequest(
    val items: List<ClaimItemDto>,
)

@Serializable
data class VersionResponse(
    val version: Int,
)

File: schism-backend/internal/api/dto.go
package api

import (
	"time"

	"github.com/schism/schism-backend/internal/store"
)

type participantDTO struct {
	ID     *string `json:"id"`
	Name   string  `json:"name"`
	UserID *string `json:"userId"`
	Phone  *string `json:"phone"`
}
type groupFormDTO struct {
	Name         string           `json:"name"`
	Information  string           `json:"information"`
	Currency     string           `json:"currency"`
	CurrencyCode string           `json:"currencyCode"`
	Participants []participantDTO `json:"participants"`
}

func (d groupFormDTO) toInput() store.GroupInput {
	parts := make([]store.ParticipantInput, len(d.Participants))
	for i, p := range d.Participants {
		parts[i] = store.ParticipantInput{ID: p.ID, Name: p.Name, UserID: p.UserID, Phone: p.Phone}
	}
	return store.GroupInput{
		Name: d.Name, Information: d.Information, Currency: d.Currency,
		CurrencyCode: d.CurrencyCode, Participants: parts,
	}
}

type paidForDTO struct {
	ParticipantID string `json:"participantId"`
	Shares        int64  `json:"shares"`
}
type expenseFormDTO struct {
	Title           string       `json:"title"`
	Amount          int64        `json:"amount"`
	CategoryID      int          `json:"categoryId"`
	// Accepts a date-only "2006-01-02" (what the clients send) or a full RFC3339 timestamp;
	// parsed leniently in toInput so a plain date never fails JSON decoding.
	ExpenseDate     string       `json:"expenseDate"`
	PaidByID        string       `json:"paidById"`
	SplitMode       string       `json:"splitMode"`
	IsReimbursement bool         `json:"isReimbursement"`
	Notes           string       `json:"notes"`
	AddedBy         string       `json:"addedBy"`
	PaidFor         []paidForDTO `json:"paidFor"`
}

type claimItemDTO struct {
	Idx         int    `json:"idx"`
	Name        string `json:"name"`
	Qty         int    `json:"qty"`
	AmountMinor int64  `json:"amountMinor"`
}

func toStoreItems(items []claimItemDTO) []store.ClaimItem {
	out := make([]store.ClaimItem, len(items))
	for i, it := range items {
		out[i] = store.ClaimItem{Idx: it.Idx, Name: it.Name, Qty: it.Qty, AmountMinor: it.AmountMinor}
	}
	return out
}

type createClaimSessionDTO struct {
	Title         string         `json:"title"`
	Currency      string         `json:"currency"`
	Items         []claimItemDTO `json:"items"`
	TaxMinor      int64          `json:"taxMinor"`
	FeesMinor     int64          `json:"feesMinor"`
	DiscountMinor int64          `json:"discountMinor"`
	RoundoffMinor int64          `json:"roundoffMinor"`
}

func (d createClaimSessionDTO) toStoreItems() []store.ClaimItem { return toStoreItems(d.Items) }

type claimWeightDTO struct {
	ItemIdx int     `json:"itemIdx"`
	Weight  float64 `json:"weight"`
}
type putClaimsDTO struct {
	ExpectedVersion int              `json:"expectedVersion"`
	Weights         []claimWeightDTO `json:"weights"`
}

type resolutionDTO struct {
	ItemIdx       int    `json:"itemIdx"`
	Mode          string `json:"mode"`
	ParticipantID string `json:"participantId"`
}
type finalizeDTO struct {
	ExpectedVersion int             `json:"expectedVersion"`
	Resolutions     []resolutionDTO `json:"resolutions"`
}

func (d finalizeDTO) toResolutions() []store.UnclaimedResolution {
	out := make([]store.UnclaimedResolution, len(d.Resolutions))
	for i, r := range d.Resolutions {
		out[i] = store.UnclaimedResolution{ItemIdx: r.ItemIdx, Mode: r.Mode, ParticipantID: r.ParticipantID}
	}
	return out
}

type setReadyDTO struct {
	Ready bool `json:"ready"`
}

type editItemsDTO struct {
	Items []claimItemDTO `json:"items"`
}

func (d editItemsDTO) toStoreItems() []store.ClaimItem { return toStoreItems(d.Items) }

func (d expenseFormDTO) toInput() store.ExpenseInput {
	pf := make([]store.PaidForInput, len(d.PaidFor))
	for i, p := range d.PaidFor {
		pf[i] = store.PaidForInput{ParticipantID: p.ParticipantID, Shares: p.Shares}
	}
	date := time.Now()
	if d.ExpenseDate != "" {
		if t, err := time.Parse("2006-01-02", d.ExpenseDate); err == nil {
			date = t
		} else if t, err := time.Parse(time.RFC3339, d.ExpenseDate); err == nil {
			date = t
		}
	}
	mode := d.SplitMode
	if mode == "" {
		mode = "EVENLY"
	}
	return store.ExpenseInput{
		Title: d.Title, Amount: d.Amount, CategoryID: d.CategoryID, ExpenseDate: date,
		PaidByID: d.PaidByID, SplitMode: mode, IsReimbursement: d.IsReimbursement,
		Notes: d.Notes, AddedBy: d.AddedBy, PaidFor: pf,
	}
}
